overlays/nodejs/sns/snsMixins.ts

```diff
@@ -0,0 +1,107 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as pulumi from "@pulumi/pulumi";
+import * as lambda from "../lambda";
+import * as topic from "./topic";
+import * as topicSubscription from "./topicSubscription";
+
+export interface TopicEvent {
+    Records: TopicRecord[];
+}
+
+export interface TopicRecord {
+    EventVersion: string;
+    EventSubscriptionArn: string;
+    EventSource: string;
+    Sns: SNSItem;
+}
+
+export interface SNSItem {
+    SignatureVersion: string;
+    Timestamp: string;
+    Signature: string;
+    SigningCertUrl: string;
+    MessageId: string;
+    Message: string;
+    MessageAttributes: { [key: string]: SNSMessageAttribute };
+    Type: string;
+    UnsubscribeUrl: string;
+    TopicArn: string;
+    Subject: string;
+}
+
+export interface SNSMessageAttribute {
+    Type: string;
+    Value: string;
+}
+
+export type TopicEventHandler = lambda.EventHandler<TopicEvent, void>;
+
+/**
+ * Arguments to control the topic subscription.  Currently empty, but still defined in case of
+ * future need.
+ */
+export type TopicEventSubscriptionArgs = { };
+
+export class TopicEventSubscription extends lambda.EventSubscription {
+    public readonly topic: topic.Topic;
+
+    /**
+     * The underlying sns object created for the subscription.
+     */
+    public readonly subscription: topicSubscription.TopicSubscription;
+
+    public constructor(
+        name: string, topic: topic.Topic, handler: TopicEventHandler,
+        args: TopicEventSubscriptionArgs, opts?: pulumi.ResourceOptions) {
+
+        super("aws:sns:TopicEventSubscription", name, { }, opts);
+
+        args = args || {};
+        this.topic = topic;
+
+        const parentOpts = { parent: this };
+        this.func = lambda.createFunctionFromEventHandler(name, handler, parentOpts);
+
+        this.permission = new lambda.Permission(name, {
+            action: "lambda:invokeFunction",
+            function: this.func,
+            principal: "sns.amazonaws.com",
+            sourceArn: topic.id,
+        }, parentOpts);
+
+        this.subscription = new topicSubscription.TopicSubscription(name, {
+            topic: topic,
+            protocol: "lambda",
+            endpoint: this.func.arn,
+        }, parentOpts);
+    }
+}
+
+declare module "./topic" {
+    interface Topic {
+        /**
+         * Creates a new subscription to events fired from this Topic to the handler provided, along
+         * with options to control the behavior of the subscription.
+         */
+        onEvent(
+            name: string, handler: TopicEventHandler,
+            args?: TopicEventSubscriptionArgs, opts?: pulumi.ResourceOptions): TopicEventSubscription;
```

Comment:
bikeshedding. how does .onEvent sound to people here?   we could say .onNewMessages, but i don't think there's high value in changing hte name.